File: Project/src/code.myceliUs.com/CargoWebServer/Cargo/Server/EntityPrototype.go
package Server

import "fmt"
import "strings"
import "log"
import "code.myceliUs.com/Utility"

/**
 * Restrictions for Datatypes
 */
type RestrictionType int

const (
	// Defines a list of acceptable values
	RestrictionType_Enumeration RestrictionType = 1 + iota

	// Specifies the maximum number of decimal places allowed. Must be equal to
	// or greater than zero
	RestrictionType_FractionDigits

	// Specifies the exact number of characters or list items allowed. Must be
	// equal to or greater than zero
	RestrictionType_Length

	// Specifies the upper bounds for numeric values (the value must be less
	// than this value)
	RestrictionType_MaxExclusive

	// Specifies the upper bounds for numeric values (the value must be less than
	// or equal to this value)
	RestrictionType_MaxInclusive

	// Specifies the maximum number of characters or list items allowed. Must be
	// equal to or greater than zero
	RestrictionType_MaxLength

	// Specifies the lower bounds for numeric values (the value must be greater
	// than this value)
	RestrictionType_MinExclusive

	// Specifies the lower bounds for numeric values (the value must be greater than or equal to this value)
	RestrictionType_MinInclusive

	// Specifies the minimum number of characters or list items allowed. Must be equal to or greater than zero
	RestrictionType_MinLength

	// Defines the exact sequence of characters that are acceptable
	RestrictionType_Pattern

	// Specifies the exact number of digits allowed. Must be greater than zero
	RestrictionType_TotalDigits

	// Specifies how white space (line feeds, tabs, spaces, and carriage returns) is handled
	RestrictionType_WhiteSpace
)

// Put constaint in a field to reduce the range of
// possibles values of a given type.
// For example an email is a string with a pattern to respect.
// so the range of string restrict by a pattern became the range of
// email.
type Restriction struct {
	TYPENAME string

	// The the of the restriction (Facet)
	Type RestrictionType
	// The value...
	Value string
}

/**
 * This structure is use to make query over the key value data store.
 */
type EntityPrototype struct {

	// Uniquely identify the entity prototype.
	UUID string

	// The name of the entity
	// The type name is compose of the package name, a comma and
	// the type name itself.
	TypeName string

	// The documentation for that entity
	Documentation string

	// True if the entity prototype is an abstrac class...
	IsAbstract bool

	// In that case the prototype define a list of given item type.
	ListOf string

	// The class derived from this entity.
	SubstitutionGroup []string

	// The list of super type, equivalent to extension.
	SuperTypeNames []string

	// Restriction of the range of possible value.
	Restrictions []*Restriction

	// The ids that compose the entity...
	Ids []string

	// The indexation of this entity
	Indexs []string

	// The list of fields of the entity
	Fields []string

	// That contain the field documentation if there is so...
	FieldsDocumentation []string

	// The list of fields type of the entity
	// ex. []string:Test.Item:Ref
	// [] means the field is an array
	// string is the format of the reference in the case
	// of type other than xsd base type.
	// Type is written like PacakageName.TypeName
	// If the field is a reference to other element, (an aggregation)
	// Ref is needed at the end. Otherwise it considere at composition.
	FieldsType []string

	// Fields visibility
	FieldsVisibility []bool

	// If the field can be nil value...
	FieldsNillable []bool

	// The order of the field, use to display in tabular form...
	FieldsOrder []int

	// The fields default value.
	FieldsDefaultValue []string

	// The prototype version.
	Version string

	// Temporary container that old information about fields modifiaction.
	FieldsToDelete []int
	FieldsToUpdate []string
}

func NewEntityPrototype() *EntityPrototype {

	prototype := new(EntityPrototype)

	// Uniquely identify entity prototype.
	prototype.UUID = Utility.RandomUUID()

	prototype.Fields = make([]string, 0)
	prototype.FieldsOrder = make([]int, 0)
	prototype.FieldsType = make([]string, 0)
	prototype.FieldsDocumentation = make([]string, 0)
	prototype.FieldsNillable = make([]bool, 0)
	prototype.FieldsVisibility = make([]bool, 0)
	prototype.SuperTypeNames = make([]string, 0)
	prototype.Restrictions = make([]*Restriction, 0)
	prototype.Indexs = make([]string, 0)
	prototype.Ids = make([]string, 0)
	prototype.FieldsDefaultValue = make([]string, 0)

	// Append the default fields at begin...
	prototype.Fields = append(prototype.Fields, "UUID")
	prototype.Ids = append(prototype.Ids, "UUID")
	prototype.FieldsOrder = append(prototype.FieldsOrder, 0)
	prototype.FieldsType = append(prototype.FieldsType, "xs.string")
	prototype.FieldsVisibility = append(prototype.FieldsVisibility, false)
	prototype.FieldsDefaultValue = append(prototype.FieldsDefaultValue, "")
	prototype.FieldsNillable = append(prototype.FieldsNillable, false)

	prototype.Fields = append(prototype.Fields, "ParentUuid")
	prototype.Indexs = append(prototype.Indexs, "ParentUuid")
	prototype.FieldsOrder = append(prototype.FieldsOrder, 1)
	prototype.FieldsType = append(prototype.FieldsType, "xs.string")
	prototype.FieldsVisibility = append(prototype.FieldsVisibility, false)
	prototype.FieldsDefaultValue = append(prototype.FieldsDefaultValue, "")
	prototype.FieldsNillable = append(prototype.FieldsNillable, true)

	prototype.Fields = append(prototype.Fields, "ParentLnk")
	prototype.FieldsOrder = append(prototype.FieldsOrder, 2)
	prototype.FieldsType = append(prototype.FieldsType, "xs.string")
	prototype.FieldsVisibility = append(prototype.FieldsVisibility, false)
	prototype.FieldsDefaultValue = append(prototype.FieldsDefaultValue, "")
	prototype.FieldsNillable = append(prototype.FieldsNillable, true)

	return prototype
}

/**
 * That function is use to create an extension of a given prototype.
 */
func (this *EntityPrototype) setSuperTypeFields() {
	var index = 3 // The start index is after the uuid and parentUuid and the parent lnk.
	if len(this.ListOf) > 0 {
		// It can not have superpetype and list of a the same time.
		this.SuperTypeNames = make([]string, 0)
		// In that particular case the entity is a list of the given type.
		if !Utility.Contains(this.Fields, "M_listOf") {
			Utility.InsertStringAt(index, "M_listOf", &this.Fields)
			Utility.InsertStringAt(index, "[]"+this.ListOf, &this.FieldsType)
			Utility.InsertBoolAt(index, true, &this.FieldsVisibility)
			Utility.InsertStringAt(index, "[]", &this.FieldsDefaultValue)
			Utility.InsertStringAt(index, "", &this.FieldsDocumentation)
		}
	} else {
		for i := 0; i < len(this.SuperTypeNames); i++ {
			superTypeName := this.SuperTypeNames[i]
			superPrototype, err := GetServer().GetEntityManager().getEntityPrototype(superTypeName, superTypeName[0:strings.Index(superTypeName, ".")])
			if err == nil {
				// I will merge the fields
				// The first to fields are always the uuid, parentUuid, parentLnk and the last is the childUuids and referenced
				for j := 3; j < len(superPrototype.Fields)-2; j++ {
					if !Utility.Contains(this.Fields, superPrototype.Fields[j]) && strings.HasPrefix(superPrototype.Fields[j], "M_") {

						Utility.InsertStringAt(index, superPrototype.Fields[j], &this.Fields)
						Utility.InsertStringAt(index, superPrototype.FieldsType[j], &this.FieldsType)
						Utility.InsertBoolAt(index, superPrototype.FieldsVisibility[j], &this.FieldsVisibility)
						Utility.InsertStringAt(index, superPrototype.FieldsDefaultValue[j], &this.FieldsDefaultValue)

						// create a new index at the end...
						if superPrototype.FieldsNillable != nil {
							isNillable := false
							if j < len(superPrototype.FieldsNillable) {
								isNillable = superPrototype.FieldsNillable[j]
							}
							Utility.InsertBoolAt(index, isNillable, &this.FieldsNillable)
						} else {
							this.FieldsNillable = append(this.FieldsNillable, true)
						}

						if superPrototype.FieldsDocumentation != nil {
							documentation := ""
							if j < len(superPrototype.FieldsDocumentation) {
								documentation = superPrototype.FieldsDocumentation[j]
							}
							if index < len(this.FieldsDocumentation) {
								Utility.InsertStringAt(index, documentation, &this.FieldsDocumentation)
							} else {
								this.FieldsDocumentation = append(this.FieldsDocumentation, documentation)
							}
						} else {
							this.FieldsDocumentation = append(this.FieldsDocumentation, "")
						}

						index++
					}
				}
				// Now the index...
				for j := 0; j < len(superPrototype.Indexs); j++ {
					if !Utility.Contains(this.Indexs, superPrototype.Indexs[j]) {
						this.Indexs = append(this.Indexs, superPrototype.Indexs[j])
					}
				}
				// Now the ids
				for j := 0; j < len(superPrototype.Ids); j++ {
					if !Utility.Contains(this.Ids, superPrototype.Ids[j]) {
						this.Ids = append(this.Ids, superPrototype.Ids[j])
					}
				}

				// Now I will append the new prototype to the list of substitution group of the super type.
				if !Utility.Contains(superPrototype.SubstitutionGroup, this.TypeName) {
					if !Utility.Contains(superPrototype.SubstitutionGroup, this.TypeName) {
						superPrototype.SubstitutionGroup = append(superPrototype.SubstitutionGroup, this.TypeName)
						// save it to it store...
						store := GetServer().GetDataManager().getDataStore(superTypeName[0:strings.Index(superTypeName, ".")])
						store.SaveEntityPrototype(superPrototype)
					}
				}
			} else {
				log.Println("error ", err)
			}
		}
	}

	// reset the field orders.
	this.FieldsOrder = make([]int, len(this.Fields))
	for i := 0; i < len(this.Fields); i++ {
		this.FieldsOrder[i] = i
	}

}

/**
 * This function is use to retreive the position in the array of a given field.
 */
func (this *EntityPrototype) getFieldIndex(fieldName string) int {

	if this.Fields != nil {
		for i := 0; i < len(this.Fields); i++ {
			if this.Fields[i] == fieldName {
				return i
			}
		}
	}
	return -1
}

/**
 * Generate the JavaScript class defefinition.
 */
func (this *EntityPrototype) generateConstructor() string {
	var packageName string

	values := strings.Split(this.TypeName, ".")
	packageName = values[0]
	var constructorSrc = "var " + packageName + " = " + packageName + "|| {};\n"

	// create sub-namspace if there is some.
	if len(values) > 2 {
		for i := 1; i < len(values)-1; i++ {
			packageName += "." + values[i]
			constructorSrc += packageName + " = " + packageName + "|| {};\n"
		}
	}

	constructorSrc += this.TypeName + " = function(){\n"

	// Common properties share by all entity.
	constructorSrc += " this.__class__ = \"" + this.TypeName + "\"\n"
	constructorSrc += " this.TYPENAME = \"" + this.TypeName + "\"\n"
	constructorSrc += " this.UUID = \"\"\n"
	constructorSrc += " this.ParentUuid = \"\"\n"
	constructorSrc += " this.ParentLnk = \"\"\n"
	constructorSrc += " this.childsUuid = []\n"
	constructorSrc += " this.references = []\n"
	constructorSrc += " this.NeedSave = true\n"
	constructorSrc += " this.IsInit = false\n"
	constructorSrc += " this.exist = false\n"
	constructorSrc += " this.initCallback = undefined\n"

	// Create field and set her initial values.
	for i := 3; i < len(this.Fields); i++ {
		if len(this.FieldsDefaultValue[i]) != 0 {
			constructorSrc += " this." + this.Fields[i] + " = " + this.FieldsDefaultValue[i] + "\n"
		} else {
			if strings.HasPrefix(this.FieldsType[i], "[]") {
				constructorSrc += " this." + this.Fields[i] + " = []\n"
			} else {
				constructorSrc += " this." + this.Fields[i] + " = null\n"
			}
		}
	}

	// Keep the reference on the entity prototype.
	constructorSrc += " return this\n"
	constructorSrc += "}\n"

	return constructorSrc
}

/**
 * For debug purpose only...
 */
func (this *EntityPrototype) Print() {
	// The prototype Type Name...
	fmt.Println("\nTypeName:", this.TypeName)
	if len(this.SuperTypeNames) > 0 {
		fmt.Println("	Super Types:", this.SuperTypeNames)
	}

	if this.SubstitutionGroup != nil {
		fmt.Println("	Substitution Groups:", this.SubstitutionGroup)
	}

	if len(this.ListOf) > 0 {
		fmt.Println("	List of:", this.ListOf)
	}

	// Now the restrictions...
	if this.Restrictions != nil {
		for j := 0; j < len(this.Restrictions); j++ {
			if this.Restrictions[j].Type == RestrictionType_Enumeration {
				fmt.Println("	----> Enumration Restriction:", this.Restrictions[j].Value)
			} else if this.Restrictions[j].Type == RestrictionType_FractionDigits {
				fmt.Println("	----> Fraction Digits Restriction:", this.Restrictions[j].Value)
			} else if this.Restrictions[j].Type == RestrictionType_Length {
				fmt.Println("	----> Length Restriction:", this.Restrictions[j].Value)
			} else if this.Restrictions[j].Type == RestrictionType_MaxExclusive {
				fmt.Println("	----> Max Exclusive Restriction:", this.Restrictions[j].Value)
			} else if this.Restrictions[j].Type == RestrictionType_MaxInclusive {
				fmt.Println("	----> Max Inclusive Restriction:", this.Restrictions[j].Value)
			} else if this.Restrictions[j].Type == RestrictionType_MaxLength {
				fmt.Println("	----> Max Length Restriction:", this.Restrictions[j].Value)
			} else if this.Restrictions[j].Type == RestrictionType_MinExclusive {
				fmt.Println("	----> Min Exclusive Restriction:", this.Restrictions[j].Value)
			} else if this.Restrictions[j].Type == RestrictionType_MinInclusive {
				fmt.Println("	----> Min Inclusive Restriction:", this.Restrictions[j].Value)
			} else if this.Restrictions[j].Type == RestrictionType_MinLength {
				fmt.Println("	----> Min Length Restriction:", this.Restrictions[j].Value)
			} else if this.Restrictions[j].Type == RestrictionType_TotalDigits {
				fmt.Println("	----> Total Digits Restriction:", this.Restrictions[j].Value)
			} else if this.Restrictions[j].Type == RestrictionType_WhiteSpace {
				fmt.Println("	----> White Space Restriction:", this.Restrictions[j].Value)
			}
		}
	}

	// Now the fields...
	fmt.Println("	Fields:")
	for i := 0; i < len(this.Fields); i++ {
		fmt.Println("	-->", this.Fields[i], ":", this.FieldsType[i])
	}
}

File: WebApp/Cargo/Script/src/Cargo/entityPrototype.js
/**
 * Global variable where the entities will be keep in memory.
 */
require("./eventHub")

 // Store the prototypes.
var entityPrototypes = {};

// Store the entities.
var entities = {};

/**
 * The entity prototype define the schema of an entity. Entity prototype
 * are usefull to make entity persistent, to export and import schema from
 * other data source like xml schema. It's also possible to create at runtime
 * new kind of entity and make query over it.
 * </br>note: Fields, FieldsType, FieldsDocumentation, FieldsNillable and FieldsOrder 
 * must have the same number of elements.
 * @constructor 
 */
var EntityPrototype = function () {
    this.TYPENAME = "Server.EntityPrototype"

    /**
     * @property {string} TypeName The type name of the entity must be write 'packageName.className'
     */
    this.TypeName = "" // string

    /** 
     * @property {string} Documentation The entity type documentation.
     */
    this.Documentation = ""

    /** 
     * @property {boolean} IsAbstract True if the entity prototype is an abstrac class.
     */
    this.IsAbstract = false

    /**
     * @property ListOf The prototype consiste of a list of other type
     */
    this.ListOf = ""

    /**
     * @property SubstitutionGroup The list of substitution group, must be use in abstract class.
     */
    this.SubstitutionGroup = []

    /**
     * @property SuperTypeNames The list of super class name.
     */
    this.SuperTypeNames = [] // string

    /**
     *  @property Restrictions The field restrictions.
     */
    this.Restrictions = [] // Restriction

    /**
     * @property Ids The list of fields used as ids.
     */
    this.Ids = [] // string

    /**
     * @property Indexs The list of fields used as index.
     */
    this.Indexs = [] // string

    ////////////////////////////////////////////////////////////////////////////
    //  Field's properties.
    ////////////////////////////////////////////////////////////////////////////

    /** 
     * @property Fields The fields name 
     */
    this.Fields = []

    /**
     * @property FieldsDocumentation The field documentation
     */
    this.FieldsDocumentation = []

    /**
     * @property FieldsType The fields type
     */
    this.FieldsType = [] // string

    /**
     * @property FieldsVisibility The fields visibility
     */
    this.FieldsVisibility = [] // bool

    /**
     * @property FieldsNillable If the field can be nil
     */
    this.FieldsNillable = [] // bool

    /**
     * @property The fields default value. Use to initialyse a new object with a given value.
     */
    this.FieldsOrder = []

    /**
     * @property FieldsOrder The field order
    */
    this.FieldsDefaultValue = [] // int

    /**
     * @property Contain the list of fields to delete
    */
    this.FieldsToDelete = [] // int

    /**
     * @property Contain the list of fields to rename
    */
    this.FieldsToUpdate = [] // int

    return this
}

/**
 * Return the list of field to be used as title.
 */
EntityPrototype.prototype.getTitles = function () {
    // The get title default function... can be overload.
    var titles = []

    // The ids
    for (var i = 1; i < this.Ids.length; i++) {
        var fieldIndex = this.getFieldIndex(this.Ids[i])
        var field = this.Fields[fieldIndex]
        if (this.FieldsVisibility[fieldIndex] == true) {
            titles.push(this[field])
        }
    }

    // The indexs
    for (var i = 1; i < this.Indexs.length; i++) {
        var fieldIndex = this.getFieldIndex(this.Indexs[i])
        var field = this.Fields[fieldIndex]
        if (this.FieldsVisibility[fieldIndex] == true) {
            titles.push(this[field])
        }
    }
    return titles
}


/**
 * Create a new class form json object.
 * @param {object} object The object that regroup the prototype properties.
 */
EntityPrototype.prototype.init = function (object) {

    if (object == null || object.TypeName == undefined) {
        return
    }

    if (object.FieldsNillable == null) {
        object.FieldsNillable = []
        for (var i = 0; i < object.Fields.length; i++) {
            object.FieldsNillable.push(false)
        }
    }

    if (object.FieldsDocumentation == null) {
        object.FieldsDocumentation = []
        for (var i = 0; i < object.Fields.length; i++) {
            object.FieldsDocumentation.push("")
        }
    }
	

    // The type name.
    this.TypeName = object.TypeName

    // The package will be an object on the global scope.
    this.PackageName = object.TypeName.split(".")[0]

    // The type name.
    this.ClassName = object.TypeName.substring(object.TypeName.indexOf(".") + 1)

    // The object ids
    this.appendIds(object.Ids)

    // The object indexs
    this.appendIndexs(object.Indexs)

    // Now the fields.
    if (object.Fields != null && object.FieldsType != null && object.FieldsVisibility != null && object.FieldsOrder != null) {
        if (!contains(object.Fields, "ParentLnk")) {
            object.Fields.unshift("ParentLnk")
            object.FieldsType.unshift("xs.string")
            object.FieldsVisibility.unshift(false)
            object.FieldsNillable.unshift(false)
            object.FieldsDocumentation.unshift("Relation with it parent.")
            object.FieldsOrder.push(object.FieldsOrder.length)
            object.FieldsDefaultValue.unshift("")
        }

        // Append parent uuid if none is define.
        if (!contains(object.Fields, "ParentUuid")) {
            object.Fields.unshift("ParentUuid")
            object.FieldsType.unshift("xs.string")
            object.FieldsVisibility.unshift(false)
            object.FieldsNillable.unshift(false)
            object.FieldsDocumentation.unshift("The parent object UUID")
            object.FieldsOrder.push(object.FieldsOrder.length)
            object.FieldsDefaultValue.unshift("")
        }

        // Append the uuid if none is define.
        if (!contains(object.Fields, "UUID")) {
            // append the uuid...
            object.Fields.unshift("UUID")
            object.FieldsType.unshift("xs.string")
            object.FieldsVisibility.unshift(false)
            object.FieldsOrder.push(object.FieldsOrder.length)
            object.FieldsNillable.unshift(false)
            object.FieldsDocumentation.unshift("The object UUID")
            object.FieldsDefaultValue.unshift("")
            object.Ids.unshift("UUID")
        }
	
        for (var i = 0; i < object.Fields.length; i++) {
            this.appendField(object.Fields[i], object.FieldsType[i], object.FieldsVisibility[i], object.FieldsOrder[i], object.FieldsNillable[i], object.FieldsDocumentation[i], object.FieldsDefaultValue[i])
            if (object.Fields[i] == "UUID") {
                if (!contains(this.Ids, "UUID")) {
                    this.Ids.unshift("UUID")
                }
            } else if (object.Fields[i] == "ParentUuid") {
                if (!contains(this.Indexs, "ParentUuid")) {
                    this.Indexs.unshift("ParentUuid")
                }
            } else if (object.Fields[i] == "ParentLnk") {
                if (!contains(this.Indexs, "ParentLnk")) {
                    this.Indexs.unshift("ParentLnk")
                }
            }
        }

    } else {
        console.log(object.TypeName + " has no fields!!!")
    }

    // If the object is abstract
    this.IsAbstract = object.IsAbstract

    // Now the restriction
    this.Restrictions = object.Restrictions


    // The list of substitution type.
    if (object.SubstitutionGroup != undefined) {
        this.SubstitutionGroup = object.SubstitutionGroup
    }

    // The list of Supertype 
    if (object.SuperTypeNames != undefined) {
        this.SuperTypeNames = object.SuperTypeNames
    }

    // if the type is a collection.
    if (object.ListOf != undefined) {
        this.ListOf = object.ListOf
    }

    if (object.FieldsDefaultValue != undefined) {
        this.FieldsDefaultValue = object.FieldsDefaultValue
    }

    // other standard fields.
    this.appendField("childsUuid", "[]xs.string", false, this.Fields.length, false, "the array of child entities.", "[]")
    this.appendField("referenced", "[]Server.EntityRef", false, this.Fields.length, false, "The field documentation.", "[]")

}

/**
 * Create a new prototypeField.
 * @param {string} name The field name.
 * @param {string} name The field type name.
 * @param {boolean} isVisible True, if the field is visible.
 * @param {int} order The order the field will be return, usefull to display.
 */
EntityPrototype.prototype.appendField = function (name, typeName, isVisible, order, isNillable, documentation, defaultValue) {
    // Set the field name.
    if (!contains(this.Fields, name)) {
        this.Fields.push(name)

        // Set the field type name
        this.FieldsType.push(typeName)

        // Set the field visibility
        this.FieldsVisibility.push(isVisible)
        // And the order (index in the list of fields.)
        this.FieldsOrder.push(parseInt(order))

        this.FieldsNillable.push(isNillable)

        this.FieldsDocumentation.push(documentation)

        this.FieldsDefaultValue.push(defaultValue)
    }
}

/**
 * Append the list of indexs
 * @param indexs The list of indexs to append.
 */
EntityPrototype.prototype.appendIndexs = function (indexs) {
    if (indexs != null) {
        for (var i = 0; i < indexs.length; i++) {
            this.Indexs.push(indexs[i])
        }
    }
}

/**
 * Append the list of id's
 * @param ids The list of ids to append.
 */
EntityPrototype.prototype.appendIds = function (ids) {
    if (ids != null) {
        for (var i = 0; i < ids.length; i++) {
            this.Ids.push(ids[i])
        }
    }
}

/**
 * Retreive the index of a given field in the prototype.
 * @param field The field we looking for.
 * @returns The index of the field or -1 if the field is not there.
 */
EntityPrototype.prototype.getFieldIndex = function (field) {
    for (var i = 0; i < this.Fields.length; i++) {
        if (this.Fields[i] == field) {
            return i
        }
    }
    return -1
}


/**
 * That class is use to manage entity prototype event.
 */
var EntityPrototypeManager = function () {
    if (server == undefined) {
        return
    }
    EventHub.call(this, PrototypeEvent)

    return this
}

EntityPrototypeManager.prototype = new EventHub(null);
EntityPrototypeManager.prototype.constructor = EntityPrototypeManager;

/**
 * Overload the onEvent function to set the new  prototype object on the local map
 * before propagate the event to listener.
 */
EntityPrototypeManager.prototype.onEvent = function (evt) {
    if (evt.code == NewPrototypeEvent || evt.code == UpdatePrototypeEvent) {
        // Set the prototype.
        var prototype = new EntityPrototype()
        prototype.init(evt.dataMap.prototype)
        setEntityPrototype(prototype)
        evt.dataMap.prototype = prototype
    } else if (evt.code == DeletePrototypeEvent) {
        // Remove it from the map
        delete entityPrototypes[evt.dataMap.prototype.TypeName]
    }

    // Call the regular function.
    EventHub.prototype.onEvent.call(this, evt);
}

/**
 * Use that function to set value of the global map.
 * @param {*} prototype 
 */
function setEntityPrototype(prototype) {
    entityPrototypes[prototype.TypeName] = prototype
}

// List of exports...
// functions
exports.setEntityPrototype = setEntityPrototype
// variables
exports.entityPrototypes = entityPrototypes
exports.entities = entities
// class
exports.EntityPrototype = EntityPrototype
exports.EntityPrototypeManager = EntityPrototypeManager

